### PR TITLE
Junebug handler

### PR DESCRIPTION
**Pull request by nicpottier**
No USSD for now support as that is still an unknown going forward.

### CODE REVIEW

**Comment by nicpottier** [comment]
@smn if you could take a look at this would be useful, we don't have any active Junebug channels on our installs yet, but want to make sure all messaging channels are supported in Courier.

As an aside, is Vumi still used anywhere? We don't have any active installs and wondering if removing it entirely from RapidPro may make sense if that has been replaced by Junebug.

File: handlers/junebug/junebug.go [hunk 1]
```diff
@@ -1,3 +1,227 @@
...
+
+	urn := urns.NewTelURNForCountry(payload.From, c.Country())
+	msg := h.Backend().NewIncomingMsg(c, urn, payload.Content).WithExternalID(payload.MessageID).WithReceivedOn(date.UTC())
+
+	err = h.Backend().WriteMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.Event{msg}, courier.WriteMsgSuccess(ctx, w, r, []courier.Msg{msg})
```
Comment (nicpottier) [inline handlers/junebug/junebug.go]: @smn one question is whether Junebug only care about status codes when delivering events and messages to RapidPro or whether it also parses the body. Right now this is returning the generic Courier JSON response bodies for incoming msgs and status updates.

**Comment by smn** [comment]
@nicpottier thanks for the ping, would love to review but on paternity atm, if this is time sensitive then perhaps @rudigiesler would be a better person to review. If not, I can review in about a week or so.

Wrt Vumi, I think it makes sense to drop support for that. All our installs going forward are around Junebug.

**Comment by nicpottier** [comment]
Congratulations @smn! 👶 

So is that to say you guys don't have any active Vumi channels on any of your RapidPro instances? If so then ya I think it makes sense to remove it.

@rudigiesler your eyes would be appreciated!

**Comment by smn** [comment]
@nicpottier thanks :) 

Yeah, afaik, none of our production rapidpro instances are connected to Vumi but @rudigiesler and @miltontony should confirm as they're more involved with those things nowadays.

**Comment by rudigiesler** [comment]
@nicpottier Yes to removing the vumi go channel. Vumi Go is officially marked as being deprecated and no longer supported, and has been scaled down to a bare minimum. We were planning to switch it off beginning of this year, but there are still a few projects hanging on to it, so we're in the process of moving those off and then we'll switch Vumi Go off.

So yes to removing Vumi Go, Junebug is the replacement for anything running on there.

File: handlers/junebug/junebug.go [hunk 2]
```diff
@@ -1,3 +1,227 @@
...
+//   "reply_to": null,
+//   "message_id": "message-id"
+// }
+type moPayload struct {
+	From      string `json:"from"       validate:"required"`
+	Timestamp string `json:"timestamp"  validate:"required"`
+	Content   string `json:"content"`
+	To        string `json:"to"         validate:"required"`
+	ReplyTo   string `json:"reply_to"`
+	MessageID string `json:"message_id" validate:"required"`
```
[the remark this answers is not in the document]
Comment (nicpottier) [reply handlers/junebug/junebug.go]: Ya, didn't see those used anywhere so we're only parsing those we need. (extra fields are fine in golang json parsing)

File: handlers/junebug/junebug.go [hunk 3]
```diff
@@ -1,3 +1,227 @@
...
+	payload := &moPayload{}
+	err := handlers.DecodeAndValidateJSON(payload, r)
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, c, err)
+	}
+
+	// check authentication
+	secret := c.StringConfigForKey(courier.ConfigSecret, "")
+	if secret != "" {
+		authorization := r.Header.Get("HTTP_AUTHORIZATION")
```
[the remark this answers is not in the document]
Comment (nicpottier) [reply handlers/junebug/junebug.go]: Hah, no this is totally just me doing a braindead translation of the Django code, which ya uses weird things like `HTTP_AUTHORIZATION` for looking up that header. Will change to `Authorization`, thanks for the catch!

File: handlers/junebug/junebug.go [hunk 4]
```diff
@@ -1,3 +1,227 @@
...
+	EventType string `json:"event_type" validate:"required"`
+	MessageID string `json:"message_id" validate:"required"`
+}
+
+var statusMapping = map[string]courier.MsgStatusValue{
+	"submitted":          courier.MsgSent,
+	"delivery_succeeded": courier.MsgDelivered,
+	"delivery_failed":    courier.MsgFailed,
+	"rejected":           courier.MsgFailed,
+}
```
[the remark this answers is not in the document]
Comment (nicpottier) [reply handlers/junebug/junebug.go]: Great, will add that.

**Comment by nicpottier** [comment]
Thanks for the look Rudi! We'll tweak this PR and merge this today. We'll plan on removing Vumi in the next week or so from RapidPro.

**Pull request merged**